src/Qir/Runtime/public/QubitManager.hpp

```diff
@@ -0,0 +1,220 @@
+    struct QubitListInSharedArray final
+    {
+    private:
+        QubitIdType firstElement = NoneMarker;
+        QubitIdType lastElement = NoneMarker;
+        // We are not storing pointer to shared array because it can be reallocated.
+        // Indexes and special values remain the same on such reallocations.
+
+    public:
+        // Initialize empty list
+        QubitListInSharedArray() = default;
+
+        // Initialize as a list with sequential elements from startId to endId inclusve.
+        QubitListInSharedArray(QubitIdType startId, QubitIdType endId, QubitIdType* sharedQubitStatusArray);
+
+        bool IsEmpty() const;
+        void AddQubit(QubitIdType id, bool addToFront, QubitIdType* sharedQubitStatusArray);
+        int TakeQubitFromFront(QubitIdType* sharedQubitStatusArray);
+        void MoveAllQubitsFrom(QubitListInSharedArray& source, QubitIdType* sharedQubitStatusArray);
+    };
+
+
+    // Restricted reuse area consists of multiple segments. Qubits released in one segment cannot be reused in another.
+    // One restricted reuse area can be nested in a segment of another restricted reuse area.
+    // This class tracks current segment of an area. Previous segments are tracked collectively (not individually).
+    // This class is small, contains no C++ pointers and relies on default shallow copying/destruction.
+    struct RestrictedReuseArea final
+    {
+    public:
+        QubitListInSharedArray FreeQubitsReuseProhibited;
+        QubitListInSharedArray FreeQubitsReuseAllowed;
+
+        RestrictedReuseArea() = default;
+        RestrictedReuseArea(QubitListInSharedArray freeQubits);
+    };
+
+
+
+    // This is NOT a stack! We modify it only by push/pop, but we also iterate over elements.
+    // TODO: Better name?
+    class CRestrictedReuseAreaStack final : public std::vector<RestrictedReuseArea>
+    {
+    public:
+        // No complex scenarios for now. Don't need to support copying/moving.
+        CRestrictedReuseAreaStack() = default;
+        CRestrictedReuseAreaStack(const CRestrictedReuseAreaStack&) = delete;
+        CRestrictedReuseAreaStack& operator = (const CRestrictedReuseAreaStack&) = delete;
+        ~CRestrictedReuseAreaStack() = default;
+
+        void PushToBack(RestrictedReuseArea area);
+        RestrictedReuseArea PopFromBack();
+        RestrictedReuseArea& PeekBack();
+        // TODO: Remove and use size directly?
+        int Count() const;
+    };
+
+    class QIR_SHARED_API CQubitManager
+    {
+    public:
+        CQubitManager(
+            QubitIdType initialQubitCapacity = DefaultQubitCapacity,
+            bool mayExtendCapacity = true,
+            bool encourageReuse = true);
+
+        // No complex scenarios for now. Don't need to support copying/moving.
+        CQubitManager(const CQubitManager&) = delete;
+        CQubitManager& operator = (const CQubitManager&) = delete;
+        ~CQubitManager();
```

Comment:
Class has at least one virtual function - `CreateQubitObject()`. Consider making the destructor also _virtual_. See [Item 7: Declare destructors virtual in polymorphic base classes](http://wavelino.coffeecup.com/pdf/Effective%20Cshapshap.pdf).

Reply:
Of course! We allow derived classes here. Thanks for catching. Updated.